File: internal/mapping/deployment_plan.go
package mapping

import (
	"github.com/distr-sh/distr/api"
	"github.com/distr-sh/distr/internal/types"
)

func DeploymentPlanToAPI(plan types.DeploymentPlan) api.DeploymentPlan {
	return api.DeploymentPlan{
		ID:                         plan.ID,
		CreatedAt:                  plan.CreatedAt,
		SealedAt:                   plan.SealedAt,
		PublishedByUserAccountID:   plan.PublishedByUserAccountID,
		ApplicationID:              plan.ApplicationID,
		ReleaseBundleID:            plan.ReleaseBundleID,
		ChannelID:                  plan.ChannelID,
		EnvironmentID:              plan.EnvironmentID,
		ProcessSnapshotID:          plan.ProcessSnapshotID,
		VariableSnapshotID:         plan.VariableSnapshotID,
		ReleaseContract:            plan.ReleaseContract,
		PlanSchema:                 plan.PlanSchema,
		DraftID:                    plan.DraftID,
		DeploymentUnitID:           plan.DeploymentUnitID,
		TargetConfigSnapshotID:     plan.TargetConfigSnapshotID,
		ProtocolVersion:            plan.ProtocolVersion,
		SupersedesDeploymentPlanID: plan.SupersedesDeploymentPlanID,
		SupersedeReason:            plan.SupersedeReason,
		Status:                     plan.Status,
		CanonicalChecksum:          plan.CanonicalChecksum,
		Targets:                    List(plan.Targets, DeploymentPlanTargetToAPI),
		TargetComponents:           List(plan.TargetComponents, DeploymentPlanTargetComponentToAPI),
		PreflightRuns:              List(plan.PreflightRuns, DeploymentPreflightRunToAPI),
		Steps:                      List(plan.Steps, DeploymentPlanStepToAPI),
		Variables:                  List(plan.Variables, DeploymentPlanVariableToAPI),
		Issues:                     List(plan.Issues, DeploymentPlanIssueToAPI),
		ResolvedRequirements:       plan.ResolvedRequirements,
		StepEdges:                  plan.StepEdges,
	}
}

func DeploymentPreflightRunToAPI(run types.DeploymentPreflightRun) api.DeploymentPreflightRun {
	return api.DeploymentPreflightRun{
		ID:                 run.ID,
		CreatedAt:          run.CreatedAt,
		DeploymentPlanID:   run.DeploymentPlanID,
		PlanChecksum:       run.PlanChecksum,
		ActorUserAccountID: run.ActorUserAccountID,
		Status:             run.Status,
		Checks:             List(run.Checks, DeploymentPreflightCheckToAPI),
	}
}

func DeploymentPreflightCheckToAPI(check types.DeploymentPreflightCheck) api.DeploymentPreflightCheck {
	return api.DeploymentPreflightCheck{
		ID:                       check.ID,
		CreatedAt:                check.CreatedAt,
		DeploymentPreflightRunID: check.DeploymentPreflightRunID,
		DeploymentPlanID:         check.DeploymentPlanID,
		DeploymentPlanTargetID:   check.DeploymentPlanTargetID,
		DeploymentTargetID:       check.DeploymentTargetID,
		TaskID:                   check.TaskID,
		Component:                check.Component,
		CheckKey:                 check.CheckKey,
		Status:                   check.Status,
		Expected:                 check.Expected,
		Actual:                   check.Actual,
		Message:                  check.Message,
		SortOrder:                check.SortOrder,
	}
}

func DeploymentPlanTargetToAPI(target types.DeploymentPlanTarget) api.DeploymentPlanTarget {
	return api.DeploymentPlanTarget{
		ID:                     target.ID,
		DeploymentTargetID:     target.DeploymentTargetID,
		Name:                   target.Name,
		Type:                   target.Type,
		Platform:               target.Platform,
		CustomerOrganizationID: target.CustomerOrganizationID,
		SortOrder:              target.SortOrder,
	}
}

func DeploymentPlanTargetComponentToAPI(
	component types.DeploymentPlanTargetComponent,
) api.DeploymentPlanTargetComponent {
	return api.DeploymentPlanTargetComponent{
		ID:                      component.ID,
		DeploymentPlanTargetID:  component.DeploymentPlanTargetID,
		DeploymentTargetID:      component.DeploymentTargetID,
		Component:               component.Component,
		Version:                 component.Version,
		Image:                   component.Image,
		Platform:                component.Platform,
		Contracts:               component.Contracts,
		ConfigChecksum:          component.ConfigChecksum,
		ExpectedStateVersion:    component.ExpectedStateVersion,
		ExpectedStateChecksum:   component.ExpectedStateChecksum,
		ExpectedReleaseBundleID: component.ExpectedReleaseBundleID,
		SortOrder:               component.SortOrder,
	}
}

func DeploymentPlanStepToAPI(step types.DeploymentPlanStep) api.DeploymentPlanStep {
	return api.DeploymentPlanStep{
		ID:                   step.ID,
		StepKey:              step.StepKey,
		Name:                 step.Name,
		ActionType:           step.ActionType,
		ActionName:           step.ActionName,
		ExecutionLocation:    step.ExecutionLocation,
		InputBindings:        step.InputBindings,
		Condition:            step.Condition,
		TargetTags:           step.TargetTags,
		FailureMode:          step.FailureMode,
		TimeoutSeconds:       step.TimeoutSeconds,
		RetryMaxAttempts:     step.RetryMaxAttempts,
		RetryIntervalSeconds: step.RetryIntervalSeconds,
		RequiredPermissions:  step.RequiredPermissions,
		SortOrder:            step.SortOrder,
		Dependencies:         step.Dependencies,
		Included:             step.Included,
		ExcludedReason:       step.ExcludedReason,
	}
}

func DeploymentPlanVariableToAPI(variable types.DeploymentPlanVariable) api.DeploymentPlanVariable {
	return api.DeploymentPlanVariable{
		ID:            variable.ID,
		VariableSetID: variable.VariableSetID,
		VariableID:    variable.VariableID,
		Key:           variable.Key,
		Type:          variable.Type,
		IsRequired:    variable.IsRequired,
		Status:        variable.Status,
		Source:        variable.Source,
		Value:         variable.Value,
		ReferenceID:   variable.ReferenceID,
		ReferenceName: variable.ReferenceName,
		Redacted:      variable.Redacted,
		Trace:         variable.Trace,
	}
}

func DeploymentPlanIssueToAPI(issue types.DeploymentPlanIssue) api.DeploymentPlanIssue {
	return api.DeploymentPlanIssue{
		ID:        issue.ID,
		Severity:  issue.Severity,
		Code:      issue.Code,
		Field:     issue.Field,
		Message:   issue.Message,
		SortOrder: issue.SortOrder,
	}
}

File: frontend/ui/src/app/types/deployment-plan.ts
import {DeploymentType} from '@distr-sh/distr-sdk';
import {ReleaseContract} from './release-bundle';
import {VariableResolutionTraceEntry, VariableType} from './variable-set';

export type DeploymentPlanStatus = 'DRAFT' | 'VALIDATING' | 'BLOCKED' | 'READY' | 'EXPIRED' | 'EXECUTED';
export type DeploymentPlanIssueSeverity = 'blocker' | 'warning';
export type DeploymentPlanVariableStatus = 'resolved' | 'unresolved';

export interface CreateDeploymentPlanRequest {
  releaseBundleId: string;
  environmentId: string;
  targetIds: string[];
}

export interface DeploymentPlanTask {
  id: string;
  deploymentPlanId: string;
  deploymentTargetId: string;
  status: 'QUEUED' | 'RUNNING' | 'SUCCEEDED' | 'FAILED' | 'CANCELED';
}

export interface DeploymentPlan {
  id: string;
  createdAt: string;
  applicationId: string;
  releaseBundleId: string;
  channelId: string;
  environmentId: string;
  processSnapshotId?: string;
  variableSnapshotId?: string;
  releaseContract?: ReleaseContract;
  planSchema: 'distr.deployment-plan/v1' | 'distr.target-deployment-plan/v2';
  draftId?: string;
  deploymentUnitId?: string;
  targetConfigSnapshotId?: string;
  protocolVersion: 'v1' | 'v2';
  supersedesDeploymentPlanId?: string;
  supersedeReason?: string;
  previousStateSourcePlanId?: string;
  status: DeploymentPlanStatus;
  canonicalChecksum: string;
  bootstrap: boolean;
  targets: DeploymentPlanTarget[];
  targetComponents: DeploymentPlanTargetComponent[];
  preflightRuns: DeploymentPreflightRun[];
  steps: DeploymentPlanStep[];
  variables: DeploymentPlanVariable[];
  issues: DeploymentPlanIssue[];
  baselines?: DeploymentPlanBaseline[];
  changes?: DeploymentPlanChangeEntry[];
  risks?: DeploymentPlanRiskEntry[];
}

export type DeploymentPlanBaselineProjection = 'verified_v2' | 'legacy_projection' | 'bootstrap';
export type DeploymentPlanChangeKind =
  | 'bootstrap'
  | 'baseline_authority'
  | 'image'
  | 'config'
  | 'provider'
  | 'schema'
  | 'topology'
  | 'source_notes'
  | 'previous_state'
  | 'planning_limit_exceeded';
export type DeploymentPlanRiskLevel = 'low' | 'medium' | 'high' | 'critical';

export interface DeploymentPlanBaseline {
  id?: string;
  createdAt?: string;
  deploymentPlanId?: string;
  componentInstanceId: string;
  componentKey: string;
  sourceDeploymentPlanId?: string;
  externalExecutionId?: string;
  observationId?: string;
  observedAt?: string;
  desiredRevision: number;
  desiredChecksum: string;
  observationChecksum: string;
  releaseBundleId?: string;
  version: string;
  image: string;
  platform: string;
  targetConfigSnapshotId?: string;
  configChecksum: string;
  providerBindingChecksum: string;
  schemaState: string;
  schemaChecksum: string;
  topologyChecksum: string;
  projection: DeploymentPlanBaselineProjection;
  authorizesV2Execution: boolean;
  bootstrap: boolean;
  actorUserAccountId?: string;
  canonicalChecksum: string;
  sortOrder: number;
}

export interface DeploymentPlanReleaseNote {
  releaseBundleId: string;
  version: string;
  publishedAt: string;
  sourceRevision: string;
  summary: string;
}

export interface DeploymentPlanChangeEntry {
  id?: string;
  createdAt?: string;
  deploymentPlanId?: string;
  componentInstanceId?: string;
  componentKey: string;
  kind: DeploymentPlanChangeKind;
  before: string;
  after: string;
  releaseNotes?: DeploymentPlanReleaseNote[];
  forwardOnly: boolean;
  actorUserAccountId?: string;
  canonicalChecksum: string;
  sortOrder: number;
}

export interface DeploymentPlanRiskEntry {
  id?: string;
  createdAt?: string;
  deploymentPlanId?: string;
  componentKey: string;
  code: string;
  level: DeploymentPlanRiskLevel;
  blocking: boolean;
  message: string;
  actorUserAccountId?: string;
  canonicalChecksum: string;
  sortOrder: number;
}

export interface CreatePreviousStateDeploymentPlanRequest {
  successfulDeploymentPlanId: string;
  reason: string;
}

export type DeploymentPreflightStatus = 'PASSED' | 'FAILED';
export type DeploymentPreflightCheckStatus = 'PASSED' | 'FAILED';

export interface DeploymentPreflightRun {
  id: string;
  createdAt: string;
  deploymentPlanId: string;
  planChecksum: string;
  actorUserAccountId?: string;
  status: DeploymentPreflightStatus;
  checks: DeploymentPreflightCheck[];
}

export interface DeploymentPreflightCheck {
  id: string;
  createdAt: string;
  deploymentPreflightRunId: string;
  deploymentPlanId: string;
  deploymentPlanTargetId?: string;
  deploymentTargetId?: string;
  taskId?: string;
  component?: string;
  checkKey: string;
  status: DeploymentPreflightCheckStatus;
  expected: Record<string, unknown>;
  actual: Record<string, unknown>;
  message: string;
  sortOrder: number;
}

export interface DeploymentPlanTarget {
  id: string;
  deploymentTargetId: string;
  name: string;
  type: DeploymentType;
  platform: 'linux/amd64' | 'linux/arm64';
  customerOrganizationId?: string;
  sortOrder: number;
}

export interface DeploymentPlanTargetComponent {
  id: string;
  deploymentPlanTargetId: string;
  deploymentTargetId: string;
  component: string;
  version: string;
  image: string;
  platform: 'linux/amd64' | 'linux/arm64';
  contracts: string[];
  configChecksum: string;
  expectedStateVersion: number;
  expectedStateChecksum: string;
  expectedReleaseBundleId?: string;
  sortOrder: number;
}

export interface DeploymentPlanStep {
  id: string;
  stepKey: string;
  name: string;
  actionType: string;
  actionName: string;
  executionLocation: string;
  inputBindings: Record<string, unknown>;
  condition: string;
  targetTags: string[];
  failureMode: string;
  timeoutSeconds: number;
  retryMaxAttempts: number;
  retryIntervalSeconds: number;
  requiredPermissions: string[];
  sortOrder: number;
  dependencies: string[];
  included: boolean;
  excludedReason?: string;
}

export interface DeploymentPlanVariable {
  id: string;
  variableSetId: string;
  variableId: string;
  key: string;
  type: VariableType;
  isRequired: boolean;
  status: DeploymentPlanVariableStatus;
  source: string;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
  redacted: boolean;
  trace: VariableResolutionTraceEntry[];
}

export interface DeploymentPlanIssue {
  id: string;
  severity: DeploymentPlanIssueSeverity;
  code: string;
  field: string;
  message: string;
  sortOrder: number;
}
